_posts/2022-07-18-azure-ad-authentication-cypress.md

```diff
@@ -0,0 +1,185 @@
+
+## 1. Acquire a token from Azure AD 🔑
+
+The first step involves making a request in the background using some of the information available on your Azure dashboard under _App Registrations_. For this we will need the **client and tenant IDs** as well as generating a **client secret**. One important change from the previous article is that we are now using a `grant_type` of `password` instead of `client_credentials` which I will discuss further in [E2E testing approach](#e2e-testing-approach). As a result, we will also need to create a test user so we can use their login credentials to run the tests.
+
+It is important to note that this test user:
+
+- Must have MFA **disabled**
+- Must **not** be a guest user
+- You must log in at least once via the UI to grant consent to the requested scopes
+
+Now, we can emulate the request a browser makes to the Azure token endpoint when a user enters their login details on the page.
+
+```javascript
+Cypress.Commands.add('login', () => {
+ cy.request({
+  method: 'POST',
+  url: `https://login.microsoftonline.com/${Cypress.env(
+   'AZURE_TENANT_ID'
+  )}/oauth2/v2.0/token`,
+  form: true,
+  body: {
+   grant_type: 'password',
+   client_id: Cypress.env('AZURE_CLIENT_ID'),
+   client_secret: Cypress.env('AZURE_CLIENT_SECRET'),
+   scope: 'openid profile email',
+   username: Cypress.env('USERNAME'),
+   password: Cypress.env('PASSWORD'),
+  },
+ }).then((response) => {
+  // defined in step 2
+  injectTokens(response);
+ });
+});
+```
+
+As you can see, there are a number of references to `Cypress.env('...')`; we can store all these values in the `env` within our `cypress.config.ts` (as shown below) but we should be careful not to commit this change as the **client secret** and **password** are secret values. Instead we should pass them through environment variables when running tests on our CI server. i.e. `CYPRESS_AZURE_CLIENT_SECRET` and `CYPRESS_PASSWORD`
+
+```json
+{
+ "env": {
+  "AZURE_CLIENT_ID": "<Azure client ID>",
+  "AZURE_TENANT_ID": "<Azure tenant ID>",
+  "AZURE_CLIENT_SECRET": "<Azure client secret>",
+  "USERNAME": "test.user@mechanicalrock.io",
+  "PASSWORD": "password1234"
+ }
+}
+```
+
+Now that we can successfully obtain a token for our test user, we can now move onto saving that token into session storage so that it gets picked up by `@azure/msal-browser`.
+
+## 2. Save the token to the session storage 💾
+
+Prior to making a login request, `@azure/msal-browser` checks the session storage for a number of values to see if a user is currently logged in and has a valid session token. If so, the user is not redirected to log in again and can simply continue using the app as expected. This step involves populating those session storage values that `@azure/msal-browser` expects so that we can convince it that we have already logged in through the browser.
+
+![MSAL browser session values](/img/blog/azure-ad-cypress/msal-browser-session.jpg){:loading="lazy"}
```

Comment:
@qt-pixels might need your help on this one because I specified the width and height of the images but the image aspect ratio wasn't kept

i.e. I specified `width=1200px height=90px` and because the layout was less than 1200px, it squished it to 740px width but kept the 90px height

Reply:
Yeah the image's don't need to be bigger than 740px, so that'll be your first change. You'll then have the intrinsic height/width to declare. Happy to help once you have the refined image.

Reply:
Okay I'll refine the images 👍